src/input_stream.rs

```diff
@@ -341,6 +351,124 @@ impl<T: IsA<InputStream> + IsA<Seekable>> io::Seek for InputStreamRead<T> {
+    Transitioning,
+    Reading {
+        pending: Pin<
+            Box<
+                dyn std::future::Future<Output = Result<(Vec<u8>, usize), (Vec<u8>, glib::Error)>>
+                    + 'static,
+            >,
+        >,
+    },
+}
+
+impl State {
+    fn into_buffer(self) -> Vec<u8> {
+        match self {
+            State::Waiting { buffer } => buffer,
+            _ => panic!("Invalid state"),
+        }
+    }
+
+    fn get_pending(
+        &mut self,
+    ) -> &mut Pin<
+        Box<
+            dyn std::future::Future<Output = Result<(Vec<u8>, usize), (Vec<u8>, glib::Error)>>
+                + 'static,
+        >,
+    > {
+        match self {
+            State::Reading { ref mut pending } => pending,
+            _ => panic!("Invalid state"),
+        }
+    }
+}
+pub struct InputStreamAsyncBufRead<T: IsA<InputStream>> {
+    stream: T,
+    state: State,
+}
+
+impl<T: IsA<InputStream>> InputStreamAsyncBufRead<T> {
+    pub fn into_input_stream(self) -> T {
+        self.stream
+    }
+
+    pub fn input_stream(&self) -> &T {
+        &self.stream
+    }
+
+    fn new(stream: T, buffer_size: usize) -> Self {
+        let buffer = vec![0; buffer_size];
+
+        Self {
+            stream,
+            state: State::Waiting { buffer },
+        }
+    }
+
+    fn set_reading(
+        &mut self,
+    ) -> &mut Pin<
+        Box<
+            dyn std::future::Future<Output = Result<(Vec<u8>, usize), (Vec<u8>, glib::Error)>>
+                + 'static,
+        >,
+    > {
+        match self.state {
+            State::Waiting { .. } => {
+                let waiting = mem::replace(&mut self.state, State::Transitioning);
+                let buffer = waiting.into_buffer();
+                let pending = self
+                    .input_stream()
+                    .read_async_future(buffer, Priority::default());
+                self.state = State::Reading { pending };
+            }
+            State::Reading { .. } => {}
+            State::Transitioning { .. } => panic!("Invalid state"),
+        };
+
+        self.state.get_pending()
+    }
+
+    fn set_waiting(&mut self, buffer: Vec<u8>) {
+        match self.state {
+            State::Reading { .. } => self.state = State::Waiting { buffer },
+            _ => panic!("Invalid state"),
+        }
+    }
+}
+
+impl<T: IsA<InputStream>> AsyncRead for InputStreamAsyncBufRead<T> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let reader = self.get_mut();
+        let pending = reader.set_reading();
+
+        match pending.poll_unpin(cx) {
+            Poll::Ready(Ok((out_buf, res))) => {
+                buf[..res].copy_from_slice(&out_buf[..res]);
+                reader.set_waiting(out_buf);
+                Poll::Ready(Ok(res))
+            }
+            Poll::Ready(Err((_, err))) => {
+                let kind = err.kind::<crate::IOErrorEnum>().unwrap();
```

Comment:
Maybe you want to set your state to something reasonable here then? Some kind of error state that immediately errors in the future?